### PR TITLE
[FLINK-5226] [table] Use correct DataSetCostFactory and improve DataSetCalc costs.

### PR DESCRIPTION
**Pull request by fhueske**
- Improved DataSetCalc costs make projections cheap and help to push them down.
- Adapted existing tests that check optimized plans

### CODE REVIEW

**Comment by KurtYoung** [comment]
Cool, LGTM

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala [hunk 1]
```diff
@@ -73,7 +75,11 @@ class DataSetCalc(
 
     val child = this.getInput
     val rowCnt = metadata.getRowCount(child)
-    val exprCnt = calcProgram.getExprCount
+
+    // compute number of non-field access expressions (computations, conditions, etc.)
+    //   we only want to account for computations, not for simple projections.
+    val exprCnt = calcProgram.getExprList.asScala.toList.count(!_.isInstanceOf[RexInputRef])
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala]: Very good point, thanks!

**Comment by fhueske** [comment]
Thanks for the reviews @KurtYoung and @beyond1920.

I will merge this PR later today.

**Comment by fhueske** [comment]
Merging

**Comment by tonycox** [comment]
@fhueske is there any sense to createcost optimization for Stream?

**Pull request closed without merging**